interface/c_binding/src/basisu_c_binding.cpp

```diff
@@ -0,0 +1,284 @@
+            // The transcoder will clear the extra bytes followed the used blocks to 0.
+            const uint32_t width = (orig_width + 3) & ~3;
+            const uint32_t height = (orig_height + 3) & ~3;
+            required_size = (std::max(8U, width) * std::max(8U, height) * 4 + 7) / 8;
+            assert(required_size >= total_blocks * bytes_per_block);
+        }
+
+        assert(required_size <= dst_size);
+
+        status = m_transcoder.transcode_image_level(
+            m_file, byteLength, image_index, level_index,
+            dst, dst_size / bytes_per_block,
+            static_cast<basist::transcoder_texture_format>(format),
+            flags);
+    }
+
+    return status;
+}
+
+#ifdef KTX_BASISU_C_BINDINGS
+
+extern "C" {
+    
+DLL_EXPORT void ktx_basisu_basis_init()
+{
+    basisu_transcoder_init();
+    
+    if (!g_pGlobal_codebook)
+        g_pGlobal_codebook = new basist::etc1_global_selector_codebook(g_global_selector_cb_size, g_global_selector_cb);
+}
+
+DLL_EXPORT basis_file* ktx_basisu_create_basis() {
+    basis_file* new_basis = new basis_file();
+    return new_basis;
+}
+    
+DLL_EXPORT bool ktx_basisu_open_basis( basis_file* basis, const uint8_t * data, size_t length ) {
+    return basis->open(data,length);
+}
+
+DLL_EXPORT void ktx_basisu_close_basis( basis_file* basis ) {
+    basis->close();
+}
+    
+DLL_EXPORT void ktx_basisu_delete_basis( basis_file* basis ) {
+    delete basis;
+}
+
+DLL_EXPORT bool ktx_basisu_getHasAlpha( basis_file* basis ) {
+    assert(basis!=nullptr);
+    return basis->getHasAlpha();
+}
+
+DLL_EXPORT uint32_t ktx_basisu_getNumImages( basis_file* basis ) {
+    return basis->getNumImages();
+}
+
+DLL_EXPORT uint32_t ktx_basisu_getNumLevels( basis_file* basis, uint32_t image_index) {
+    return basis->getNumLevels(image_index);
+}
+
+DLL_EXPORT uint32_t ktx_basisu_getImageWidth( basis_file* basis, uint32_t image_index, uint32_t level_index) {
+    return basis->getImageWidth(image_index,level_index);
+}
+
+DLL_EXPORT uint32_t ktx_basisu_getImageHeight( basis_file* basis, uint32_t image_index, uint32_t level_index) {
+    return basis->getImageHeight(image_index,level_index);
+}
+
+DLL_EXPORT uint32_t ktx_basisu_getImageTranscodedSizeInBytes( basis_file* basis, uint32_t image_index, uint32_t level_index, uint32_t format) {
+    return basis->getImageTranscodedSizeInBytes(image_index,level_index,format);
+}
+
+DLL_EXPORT bool ktx_basisu_startTranscoding( basis_file* basis ) {
+    return basis->startTranscoding();
+}
+
+DLL_EXPORT bool ktx_basisu_transcodeImage( basis_file* basis, void* dst, size_t dst_size, uint32_t image_index, uint32_t level_index, uint32_t format, uint32_t pvrtc_wrap_addressing, uint32_t get_alpha_for_opaque_formats) {
+    return basis->transcodeImage(dst,dst_size,image_index,level_index,format,pvrtc_wrap_addressing,get_alpha_for_opaque_formats);
+}
+}
+#endif
```

Comment:
Please fix the following "errors" (I have warnings as errors turned on so these stopped my build). These are from Xcode. I haven't tried any other compilers yet.

```
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:59:52: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    if (!m_transcoder.get_image_level_info(m_file, byteLength, li, 0, 0))
                      ~~~~~~~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:70:50: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    return m_transcoder.get_total_images(m_file, byteLength);
                        ~~~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:79:46: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    if (!m_transcoder.get_image_info(m_file, byteLength, ii, image_index))
                      ~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:91:52: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    if (!m_transcoder.get_image_level_desc(m_file, byteLength, image_index, level_index, orig_width, orig_height, total_blocks))
                      ~~~~~~~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:103:52: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    if (!m_transcoder.get_image_level_desc(m_file, byteLength, image_index, level_index, orig_width, orig_height, total_blocks))
                      ~~~~~~~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:118:52: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    if (!m_transcoder.get_image_level_desc(m_file, byteLength, image_index, level_index, orig_width, orig_height, total_blocks))
                      ~~~~~~~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:155:51: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    return m_transcoder.start_transcoding(m_file, byteLength);
                        ~~~~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:169:52: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    if (!m_transcoder.get_image_level_desc(m_file, byteLength, image_index, level_index, orig_width, orig_height, total_blocks))
                      ~~~~~~~~~~~~~~~~~~~~         ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:185:21: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
            m_file, byteLength, image_index, level_index,
                    ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:213:21: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
            m_file, byteLength, image_index, level_index,
                    ^~~~~~~~~~
/Users/mark/Projects/github/KTX-Software/interface/c_binding/src/basisu_c_binding.cpp:214:27: error: implicit conversion loses integer precision: 'unsigned long' to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
            dst, dst_size / bytes_per_block,
                 ~~~~~~~~~^~~~~~~~~~~~~~~~~
```